### PR TITLE
add: TaggingService functionality to Lambda Provider

### PR DESCRIPTION
**Pull request by gregfurman**
## Motivation
- This PR adds tagging functionality to the Lambda provider that leverages the TaggingService. In addition, we add tagging functionality for event source mapping. 
- Fixes https://github.com/localstack/localstack/issues/11739

## Changes
- Extended `TestLambdaTag` suite to include tagging tests for Event Source Mappings
- Refactored/reworked the internal `tag_resource`, `list_tags`, and `untag_resource` Lambda provider functions to use a provider-wde TaggingService. This allows for tagging of Lambda Event Source Mapping or Lambda Function resources (with support for CSC but this is not yet implemented).

## Testing
- Manually verified event source mapping creation with Terraform, using the [terraform-aws-lambda/event-source-mapping](https://github.com/terraform-aws-modules/terraform-aws-lambda/tree/master/examples/event-source-mapping) example.

### CODE REVIEW

File: localstack-core/localstack/services/lambda_/invocation/models.py [hunk 1]
```diff
@@ -16,5 +17,8 @@ class LambdaStore(BaseStore):
     # maps layer names to Layers
     layers: dict[str, Layer] = LocalAttribute(default=dict)
 
+    # maps resource ARNs for EventSourceMappings and CodeSigningConfiguration to tags
```
Comment (gregfurman) [inline localstack-core/localstack/services/lambda_/invocation/models.py:20]: This should include Lambda Functions but I'm not sure it's worth re-triggering the workflow for. Also tags are currently doubly stored both in a Lambda `Function` objects `tag` field as well as in the `TaggingService` -- with the former likely being removed in a later release.

File: localstack-core/localstack/services/lambda_/provider.py [hunk 2]
```diff
@@ -4118,87 +4141,141 @@ def delete_function_concurrency(
...
-            function.versions["$LATEST"] = dataclasses.replace(
-                latest_version, config=dataclasses.replace(latest_version.config)
-            )
 
-    def _update_tags(self, function: Function, tags: dict[str, str]):
-        with function.lock:
-            stored_tags = function.tags or {}
-            stored_tags |= tags
-            self._store_tags(function=function, tags=stored_tags)
+        tag_svc_adapted_tags = [{"Key": key, "Value": value} for key, value in tags.items()]
```
Comment (gregfurman) [inline localstack-core/localstack/services/lambda_/provider.py:4161]: We need to adapt the tags `TaggingService` <-> `LambdaProvider` when doing storage and retrieval of tags.

File: localstack-core/localstack/services/lambda_/provider.py [hunk 3]
```diff
@@ -4118,87 +4141,141 @@ def delete_function_concurrency(
...
-                Type="User",
+                "An error occurred and the request cannot be processed.", Type="User"
             )
+        self._store_tags(resource, tags)
 
-        account_id, region = api_utils.get_account_and_region(resource, context)
-        fn = self._get_function(function_name=fn_name, account_id=account_id, region=region)
-
-        self._update_tags(fn, tags)
+        if (resource_id := extract_resource_from_arn(resource)) and resource_id.startswith(
```
Comment (gregfurman) [inline localstack-core/localstack/services/lambda_/provider.py:4234]: Unfortunately we need to run these dirty checks to ensure a Revision ID gets updated

**Review by joe4dev — COMMENTED** [review]
Great tagging test additions 💯 

As discussed, it might be worth unifying the tags to avoid confusion and improve maintainability 🚀

File: localstack-core/localstack/services/lambda_/provider.py (hunk 2 above)
Comment (gregfurman) [reply localstack-core/localstack/services/lambda_/provider.py:4161]: I like the unified approach to tagging. Even though adapting in this way is a bit janky, I think the benefits of a single svc outweighs the costs of adapting to suit it.

In future, maybe we can extend the tag svc to better account for cases like this.

File: localstack-core/localstack/services/lambda_/api_utils.py [hunk 4]
```diff
@@ -108,6 +108,8 @@
 # pattern therefore we can sub this value in when appropriate.
 ARN_NAME_PATTERN_VALIDATION_TEMPLATE = "(arn:(aws[a-zA-Z-]*)?:lambda:)?([a-z]{{2}}((-gov)|(-iso([a-z]?)))?-[a-z]+-\\d{{1}}:)?(\\d{{12}}:)?(function:)?([a-zA-Z0-9-_{0}]+)(:(\\$LATEST|[a-zA-Z0-9-_]+))?"
 
+TAGGABLE_RESOURCE_ARN_PATTERN = "arn:(aws[a-zA-Z-]*):lambda:[a-z]{2}((-gov)|(-iso([a-z]?)))?-[a-z]+-\\d{1}:\\d{12}:(function:[a-zA-Z0-9-_]+(:(\\$LATEST|[a-zA-Z0-9-_]+))?|layer:([a-zA-Z0-9-_]+)|code-signing-config:csc-[a-z0-9]{17}|event-source-mapping:[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})"
```
[the remark this answers is not in the document]
Comment (gregfurman) [reply localstack-core/localstack/services/lambda_/api_utils.py:112]: Added!

File: localstack-core/localstack/services/lambda_/provider.py (hunk 3 above)
Comment (joe4dev) [reply localstack-core/localstack/services/lambda_/provider.py:4230]: +1, good idea

File: localstack-core/localstack/services/lambda_/invocation/models.py (hunk 1 above)
Comment (joe4dev) [reply localstack-core/localstack/services/lambda_/invocation/models.py:20]: Thanks for unifying the double storage 👏

File: localstack-core/localstack/services/lambda_/invocation/lambda_models.py [hunk 5]
```diff
@@ -595,7 +595,6 @@ class Function:
     provisioned_concurrency_configs: dict[str, ProvisionedConcurrencyConfiguration] = (
         dataclasses.field(default_factory=dict)
     )
-    tags: dict[str, str] | None = None
```
Comment (joe4dev) [inline localstack-core/localstack/services/lambda_/invocation/lambda_models.py:598]: 🧹

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 6]
```diff
@@ -5442,6 +5556,21 @@ def test_tag_limits(self, create_lambda_function, snapshot, aws_client):
             aws_client.lambda_.tag_resource(Resource=function_arn, Tags={"a_key": "a_value"})
         snapshot.match("tag_lambda_too_many_tags_additional", e.value.response)
 
+        # add too many tags on a CreateFunction
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.py:5559]: nice edge case testing 💯

**Review by joe4dev — APPROVED** [review]
Great work @gregfurman 👏 

Thank you for the comprehensive test coverage and unifying tag storage ✨ 


sidenote: the PR description mentions "Fixes issue#x", which auto-closes the issue. We typically ask the customer for feedback using the response-required tag and stale bot to foster feedback instead of directly closing upon merge. It also avoids notifying the customer too early before an ext image is shipped
